std/container/array.d

```diff
@@ -313,14 +313,18 @@ if (!is(Unqual!T == bool))
             }
             // enlarge
             immutable startEmplace = length;
-            import core.checkedint : mulu;
-            bool overflow;
-            const nbytes = mulu(newLength, T.sizeof, overflow);
-            if (overflow) assert(0);
-            _payload = (cast(T*) realloc(_payload.ptr,
-                            nbytes))[0 .. newLength];
+            if (_capacity < newLength)
+            {
+                import core.checkedint : mulu;
+
+                bool overflow;
+                const nbytes = mulu(newLength, T.sizeof, overflow);
+                if (overflow)
+                    assert(0);
+                _payload = (cast(T*) realloc(_payload.ptr, nbytes))[0 .. newLength];
+                _capacity = newLength;
+            }
             initializeAll(_payload.ptr[startEmplace .. newLength]);
```

Comment:
if `_capacity > newLength` then you never set the payload to the new length

I would add some more extensive tests for this function